File: app/src/main/java/com/gmail/cristiandeives/motofretado/ActivityDetectionService.java
package com.gmail.cristiandeives.motofretado;

import android.app.PendingIntent;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.os.IBinder;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import android.support.annotation.MainThread;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.UiThread;
import android.util.Log;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.location.ActivityRecognition;

public class ActivityDetectionService extends Service
        implements GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener {
    static final String EXTRA_MESSENGER = "messenger";

    private static final String TAG = ActivityDetectionService.class.getSimpleName();
    private static final int ACTIVITY_CHANGED_REQUEST_CODE = 1;
    private static final int DETECTION_INTERVAL = 5000; // ms

    private GoogleApiClient mGoogleApiClient;
    private PendingIntent mReceiverPendingIntent;
    private BroadcastReceiver mResultReceiver;
    private IntentFilter mReceiverIntentFilter;
    private Messenger mMessenger;

    @UiThread
    private void requestActivityUpdates() {
        Log.d(TAG, "registering BroadcastReceiver");
        registerReceiver(mResultReceiver, mReceiverIntentFilter);
        Log.d(TAG, "subscribing to activity updates");
        ActivityRecognition.ActivityRecognitionApi.requestActivityUpdates(mGoogleApiClient,
                DETECTION_INTERVAL, mReceiverPendingIntent);
    }

    @UiThread
    private void removeActivityUpdates() {
        Log.d(TAG, "unsubscribing from activity updates");
        ActivityRecognition.ActivityRecognitionApi.removeActivityUpdates(mGoogleApiClient,
                mReceiverPendingIntent);
        Log.d(TAG, "unregistering BroadcastReceiver");
        unregisterReceiver(mResultReceiver);
    }

    @UiThread
    private void sendMessage(int what) {
        Message msg = Message.obtain(null, what);
        try {
            mMessenger.send(msg);
        } catch (RemoteException ex) {
            Log.e(TAG, "error sending message to Handler", ex);
        }
    }

    @Override
    @MainThread
    public void onCreate() {
        Log.v(TAG, "> onCreate()");

        mGoogleApiClient = new GoogleApiClient.Builder(this)
                .addConnectionCallbacks(this)
                .addOnConnectionFailedListener(this)
                .addApi(ActivityRecognition.API)
                .build();

        mResultReceiver = new ActivityDetectionResultReceiver();
        mReceiverIntentFilter = new IntentFilter(ActivityDetectionResultReceiver.RESULT_ACTION);

        Log.v(TAG, "< onCreate()");
    }

    @Override
    @MainThread
    public int onStartCommand(@NonNull Intent intent, int flags, int startId) {
        Log.v(TAG, "> onStartCommand(intent=" + intent + ", flags=" + flags + ", startId=" + startId + ")");

        Bundle extras = intent.getExtras();
        mMessenger = (Messenger) extras.get(EXTRA_MESSENGER);

        Intent receiverIntent = new Intent(ActivityDetectionResultReceiver.RESULT_ACTION);
        receiverIntent.putExtra(ActivityDetectionResultReceiver.EXTRA_MESSENGER, mMessenger);
        mReceiverPendingIntent = PendingIntent.getBroadcast(this, ACTIVITY_CHANGED_REQUEST_CODE,
                receiverIntent, PendingIntent.FLAG_UPDATE_CURRENT);

        Log.d(TAG, "connecting to Google Play Services");
        mGoogleApiClient.connect();

        int ret = START_REDELIVER_INTENT;

        Log.v(TAG, "< onStartCommand(intent=" + intent + ", flags=" + flags + ", startId=" + startId + "): " + ret);
        return ret;
    }

    @Override
    @MainThread
    public void onDestroy() {
        Log.v(TAG, "> onDestroy()");

        removeActivityUpdates();

        Log.d(TAG, "disconnecting from Google Play Services");
        mGoogleApiClient.disconnect();
        sendMessage(TrackBusPresenter.MyHandler.MSG_ACTIVITY_DETECTION_SERVICE_DISCONNECTED);

        Log.v(TAG, "< onDestroy()");
    }

    @Override
    @MainThread
    public IBinder onBind(Intent intent) {
        Log.v(TAG, "onBind(intent=" + intent + "): null");
        return null;
    }

    @Override
    @MainThread
    public void onConnected(@Nullable Bundle connectionHint) {
        Log.v(TAG, "> onConnected(connectionHint=" + connectionHint + ")");

        requestActivityUpdates();
        sendMessage(TrackBusPresenter.MyHandler.MSG_ACTIVITY_DETECTION_SERVICE_CONNECTED);

        Log.v(TAG, "< onConnected(connectionHint=" + connectionHint + ")");
    }

    @Override
    @MainThread
    public void onConnectionSuspended(int cause) {
        Log.v(TAG, "> onConnectionSuspended(cause=" + cause + ")");

        removeActivityUpdates();
        sendMessage(TrackBusPresenter.MyHandler.MSG_ACTIVITY_DETECTION_SERVICE_DISCONNECTED);

        Log.v(TAG, "< onConnectionSuspended(cause=" + cause + ")");
    }

    @Override
    @MainThread
    public void onConnectionFailed(@NonNull ConnectionResult result) {
        Log.v(TAG, "> onConnectionFailed(result=" + result + ")");

        Log.e(TAG, "Google Play Services connection failed: " + result.getErrorMessage());
        sendMessage(TrackBusPresenter.MyHandler.MSG_GMS_CONNECTION_FAILED);
        stopSelf();

        Log.v(TAG, "< onConnectionFailed(result=" + result + ")");
    }
}

File: app/src/main/java/com/gmail/cristiandeives/motofretado/ActivityDetectionService.kt
package com.gmail.cristiandeives.motofretado

import android.app.PendingIntent
import android.app.Service
import android.content.BroadcastReceiver
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.os.IBinder
import android.os.Messenger
import android.support.annotation.MainThread
import android.support.annotation.UiThread
import android.util.Log
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.location.ActivityRecognition

@MainThread
internal class ActivityDetectionService : Service(), GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener {
    companion object {
        internal const val EXTRA_MESSENGER = "messenger"

        private val TAG = ActivityDetectionService::class.java.simpleName
        private const val ACTIVITY_CHANGED_REQUEST_CODE = 1
        private const val DETECTION_INTERVAL = 5000L // ms
    }

    private lateinit var mGoogleApiClient: GoogleApiClient
    private lateinit var mResultReceiver: BroadcastReceiver
    private lateinit var mReceiverIntentFilter: IntentFilter
    private lateinit var mMessenger: Messenger
    private var mReceiverPendingIntent: PendingIntent? = null

    override fun onCreate() {
        Log.v(TAG, "> onCreate()")

        mGoogleApiClient = GoogleApiClient.Builder(this)
                .addConnectionCallbacks(this)
                .addOnConnectionFailedListener(this)
                .addApi(ActivityRecognition.API)
                .build()

        mResultReceiver = ActivityDetectionResultReceiver()
        mReceiverIntentFilter = IntentFilter(ActivityDetectionResultReceiver.RESULT_ACTION)

        Log.v(TAG, "< onCreate()")
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        Log.v(TAG, "> onStartCommand(intent=$intent, flags=$flags, startId=$startId)")

        val messenger = intent.extras?.get(EXTRA_MESSENGER)
        if (messenger is Messenger) {
            mMessenger = messenger
            val receiverIntent = Intent(ActivityDetectionResultReceiver.RESULT_ACTION)
            receiverIntent.putExtra(ActivityDetectionResultReceiver.EXTRA_MESSENGER, mMessenger)
            mReceiverPendingIntent = PendingIntent.getBroadcast(
                    this,
                    ACTIVITY_CHANGED_REQUEST_CODE,
                    receiverIntent,
                    PendingIntent.FLAG_UPDATE_CURRENT
            )

            Log.d(TAG, "connecting to Google Play Services")
            mGoogleApiClient.connect()
        } else {
            Log.w(TAG, "could not find a Messenger in the Intent")
        }

        val ret = START_REDELIVER_INTENT

        Log.v(TAG, "< onStartCommand(intent=$intent, flags=$flags, startId=$startId): $ret")
        return ret
    }

    override fun onDestroy() {
        Log.v(TAG, "> onDestroy()")

        removeActivityUpdates()

        Log.d(TAG, "disconnecting from Google Play Services")
        mGoogleApiClient.disconnect()

        mMessenger.sendMessage(TrackBusPresenter.MyHandler.MSG_ACTIVITY_DETECTION_SERVICE_DISCONNECTED)

        Log.v(TAG, "< onDestroy()")
    }

    override fun onBind(intent: Intent): IBinder? {
        Log.v(TAG, "onBind(intent=$intent): null")
        return null
    }

    override fun onConnected(connectionHint: Bundle?) {
        Log.v(TAG, "> onConnected(connectionHint=$connectionHint)")

        requestActivityUpdates()
        mMessenger.sendMessage(TrackBusPresenter.MyHandler.MSG_ACTIVITY_DETECTION_SERVICE_CONNECTED)

        Log.v(TAG, "< onConnected(connectionHint=$connectionHint)")
    }

    override fun onConnectionSuspended(cause: Int) {
        Log.v(TAG, "> onConnectionSuspended(cause=$cause)")

        removeActivityUpdates()
        mMessenger.sendMessage(TrackBusPresenter.MyHandler.MSG_ACTIVITY_DETECTION_SERVICE_DISCONNECTED)

        Log.v(TAG, "< onConnectionSuspended(cause=$cause)")
    }

    override fun onConnectionFailed(result: ConnectionResult) {
        Log.v(TAG, "> onConnectionFailed(result=$result)")

        Log.e(TAG, "Google Play Services connection failed: ${result.errorMessage}")
        mMessenger.sendMessage(TrackBusPresenter.MyHandler.MSG_GMS_CONNECTION_FAILED)
        stopSelf()

        Log.v(TAG, "< onConnectionFailed(result=$result)")
    }

    @UiThread
    private fun requestActivityUpdates() {
        Log.d(TAG, "registering BroadcastReceiver")
        registerReceiver(mResultReceiver, mReceiverIntentFilter)

        Log.d(TAG, "subscribing to activity updates")
        ActivityRecognition.ActivityRecognitionApi.requestActivityUpdates(mGoogleApiClient,
                DETECTION_INTERVAL, mReceiverPendingIntent)
    }

    @UiThread
    private fun removeActivityUpdates() {
        Log.d(TAG, "unsubscribing from activity updates")
        if (mGoogleApiClient.isConnected) {
            ActivityRecognition.ActivityRecognitionApi.removeActivityUpdates(mGoogleApiClient,
                    mReceiverPendingIntent)

            Log.d(TAG, "unregistering BroadcastReceiver")
            unregisterReceiver(mResultReceiver)
        }
    }
}